aptos-move/move-examples/data_structures/sources/smart_vector.move

```diff
@@ -0,0 +1,461 @@
+                max(1024 /* free_write_quota */ / estimated_avg_size, 1)
+            };
+            vector::push_back(&mut v.big_vec, big_vector::empty(bucket_size));
+        };
+        big_vector::push_back(vector::borrow_mut(&mut v.big_vec, 0), val);
+    }
+
+    /// Pop an element from the end of vector `v`. It doesn't shrink the buckets even if they're empty.
+    /// Call `shrink_to_fit` explicity to deallocate empty buckets.
+    /// Aborts if `v` is empty.
+    public fun pop_back<T>(v: &mut SmartVector<T>): T {
+        assert!(!is_empty(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let big_vec_wrapper = &mut v.big_vec;
+        if (vector::length(big_vec_wrapper) != 0) {
+            let big_vec = vector::pop_back(big_vec_wrapper);
+            let val = big_vector::pop_back(&mut big_vec);
+            if (big_vector::is_empty(&big_vec)) {
+                big_vector::destroy_empty(big_vec)
+            } else {
+                vector::push_back(big_vec_wrapper, big_vec);
+            };
+            val
+        } else {
+            vector::pop_back(&mut v.inline_vec)
+        }
+    }
+
+    /// Remove the element at index i in the vector v and return the owned value that was previously stored at i in v.
+    /// All elements occurring at indices greater than i will be shifted down by 1. Will abort if i is out of bounds.
+    /// disclaimer: this function may be costly. Use it at your own discretion.
+    public fun remove<T>(v: &mut SmartVector<T>, i: u64): T {
+        let len = length(v);
+        assert!(i < len, error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let inline_len = vector::length(&v.inline_vec);
+        if (i < inline_len) {
+            vector::remove(&mut v.inline_vec, i)
+        } else {
+            big_vector::remove(vector::borrow_mut(&mut v.big_vec, 0), i - inline_len)
+        }
+    }
+
+    /// Swap the `i`th element of the vector `v` with the last element and then pop the vector.
+    /// This is O(1), but does not preserve ordering of elements in the vector.
+    /// Aborts if `i` is out of bounds.
+    public fun swap_remove<T>(v: &mut SmartVector<T>, i: u64): T {
+        let len = length(v);
+        assert!(i < len, error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let inline_len = vector::length(&v.inline_vec);
+        let big_vec_wrapper = &mut v.big_vec;
+        let inline_vec = &mut v.inline_vec;
+        if (i >= inline_len) {
+            let big_vec = vector::pop_back(big_vec_wrapper);
+            let val = big_vector::swap_remove(&mut big_vec, i - inline_len);
+            if (big_vector::is_empty(&big_vec)) {
+                big_vector::destroy_empty(big_vec)
+            } else {
+                vector::push_back(big_vec_wrapper, big_vec);
+            };
+            val
+        } else {
+            let val = vector::swap_remove(inline_vec, i);
+            if (inline_len < len) {
```

Comment:
not sure if it matters, but this branch can be one less swap.. (do not swap_remove first, but push_back to inline and then swap_remove)

Reply:
Smart!